packages/gatsby/cache-dir/dev-loader.js

```diff
@@ -142,9 +75,68 @@ class DevLoader extends BaseLoader {
   }
 
   doPrefetch(pagePath) {
-    return super.doPrefetch(pagePath).then(result => {
-      return result.payload
-    })
+    return super.doPrefetch(pagePath).then(result => result.payload)
+  }
+
+  handleStaticQueryResultHotUpdate(msg) {
+    const newResult = msg.payload.result
+
+    const cacheKey = msg.payload.id
+    const cachedResult = this.staticQueryDb[cacheKey]
+    if (!isEqual(newResult, cachedResult)) {
+      this.staticQueryDb[cacheKey] = newResult
+      ___emitter.emit(`staticQueryResult`, newResult)
+    }
+  }
+
+  handlePageQueryResultHotUpdate(msg) {
+    const newPageData = msg.payload.result
+
+    const pageDataDbCacheKey = normalizePagePath(msg.payload.id)
+    const cachedPageData = this.pageDataDb.get(pageDataDbCacheKey)?.payload
+
+    if (!isEqual(newPageData, cachedPageData)) {
+      // always update canonical key for pageDataDb
+      this.pageDataDb.set(pageDataDbCacheKey, {
+        pagePath: pageDataDbCacheKey,
+        payload: newPageData,
+        status: `success`,
+      })
+
+      const cachedPage = this.pageDb.get(pageDataDbCacheKey)
+      if (cachedPage) {
+        this.pageDb.set(
+          pageDataDbCacheKey,
+          mergePageEntry(cachedPage, newPageData)
+        )
+      }
+
+      // Additionally if those are query results for "/404.html"
+      // we have to update all paths user wanted to visit, but didn't have
+      // page for it, because we do store them under (normalized) path
+      // user wanted to visit
+      if (pageDataDbCacheKey === `/404.html`) {
+        this.notFoundPagePathsInCaches.forEach(notFoundPath => {
+          const previousPageDataEntry = this.pageDataDb.get(notFoundPath)
+          if (previousPageDataEntry) {
+            this.pageDataDb.set(notFoundPath, {
+              ...previousPageDataEntry,
+              payload: newPageData,
+            })
+          }
+
+          const previousPageEntry = this.pageDb.get(notFoundPath)
+          if (previousPageEntry) {
+            this.pageDb.set(
+              notFoundPath,
+              mergePageEntry(previousPageEntry, newPageData)
+            )
+          }
+        })
+      }
```

Comment:
This is actually really sad that we have to do this but 👍

Reply:
This can be furthered refactored, so instead of putting `notFound` data in `pageDb` / `pageDataDb`, we would just keep list of `notFound` pages and then use `/404.html` in `loader.js` logic, but I didn't want to impact production code logic in this PR (and make even more changes)